rules/Proj0039.md

```diff
@@ -0,0 +1,58 @@
+---
+parent: General
+ancestor: MSBuild
+---
+
+# Proj0039: Treat all warnings as errors is considered a bad practice
+
+MSBuild can be configured to [threat all warnings as errors]( https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-options/errors-warnings#treatwarningsaserrors).
+How tempting this may feel (ideally a solution should not contain any warnings), there are some objections to consider. Because such a policy might be counterproductive.
+
+## 1. We�re not developing in a perfect world
+When a warning pops up, the developer facing it, should solve the issue. That why we have static code analysis enabled after all. The fact that this is not happening hints that we do not live in a perfect world. By changing the severity to error for all rules, the world we live in will not change.
+
+A developer that was not willing to resolve a warning is most likely also not willing to resolve an error beyond the bare minimum: that means by suppressing the issue, disabling the rule, or by adjusting the code just to satisfy the rule, none of these resolutions is guaranteed to be an improvement.
+
+Developers that are not capable to solve a warning (correctly) are still not able to do so once its become an error. If they did find the support to solve it correctly, what would make they would do this now? Or will they come up with the solutions equal to the unwilling developer?
+
+## 2. Deprecated code
+A specific, but common scenario is when a new version of a package is released, suddenly some code has been deprecated. Obviously the code has to be adjusted once, but it is not said that the moment you need the new version of the package is also the moment to refactor all code that relies on this now deprecated code.
+
+## 3. New rules
+The introduction of new rules (both because introduced by an existing analyser or because a new analyser was introduced) will potentially report a lot of issues. They might be valid points, but no reason to halt the release of all features until all issues are solved. By (temporarily) disabling such rules, newly added code will not yet benefit from this rule.
+
+## 4. Warnings are not errors
+It is possible to create static code analysis with an error severity. The reason that this is not done by the maker of the rule is with a reason. Apparently the code is not wrong per se, but you are warned that is beneficial to change the code.
```

Comment:
```suggestion
It is possible to create static code analysis with an error severity. The reason that this is not done by the maker of the rule is with a reason. It might be inefficient, deprecated or unreadable code, but not necessarily wrong.
```